src/number_theory/primorial.lean

```diff
@@ -0,0 +1,190 @@
+    exact select_from_sum (choose (2 * n + 1)) (range (n + 1)) n (range_contains_upper_bound n),
+  simpa [sum_range_choose_halfway n] using t
+end
+
+lemma dvd_choose_of_middling_prime (p : ℕ) (is_prime : prime p) (m : ℕ)
+  (p_big : m + 1 < p) (p_small : p ≤ 2 * m + 1) : p ∣ choose (2 * m + 1) (m + 1) :=
+begin
+  have m_size : m + 1 ≤ 2 * m + 1, by exact le_of_lt (lt_of_lt_of_le p_big p_small),
+  let e := @choose_mul_fact_mul_fact (2 * m + 1) (m + 1) m_size,
+  have r : p ∣ nat.fact (2 * m + 1), exact (prime.dvd_fact is_prime).mpr p_small,
+  have s : (p ∣ nat.fact (m + 1)) → false,
+    { intros,
+      let e := (prime.dvd_fact is_prime).mp a,
+      linarith, },
+  have t : (p ∣ nat.fact (2 * m + 1 - (m + 1))) → false,
+    { intros,
+      let f := (prime.dvd_fact is_prime).mp a,
+      have t : 2 * m + 1 - (m + 1) = m, { norm_num, rw two_mul m, exact nat.add_sub_cancel m m, },
+      rw t at f,
+      clear t a s r p_small is_prime e m_size,
+      cases lt_trichotomy p m,
+      { have r : m < m + 1, exact lt_add_one m, linarith, },
+      { cases h,
+        { rw h at p_big,
+          linarith, },
+        linarith, }, },
+  rw [←e, mul_assoc] at r,
+  cases (prime.dvd_mul is_prime).1 r,
+  { exact h, },
+  cases (prime.dvd_mul is_prime).1 h,
+  cc, cc,
+end
+
+lemma prod_primes_dvd {s : finset ℕ} : ∀ (n : ℕ) (h : ∀ a ∈ s, prime a) (div : ∀ a ∈ s, a ∣ n),
+  (∏ p in s, p) ∣ n :=
+begin
+  apply finset.induction_on s,
+  { simp, },
+  { intros a s a_not_in_s induct n primes divs,
+    rw finset.prod_insert a_not_in_s,
+    have a_div_n : a ∣ n, by exact divs a (finset.mem_insert_self a s),
+    rcases a_div_n with ⟨ k, k_times ⟩,
+    rw k_times,
+    have step : ∏ p in s, p ∣ k,
+      { apply induct k,
+        { intros b b_in_s,
+          exact primes b (finset.mem_insert_of_mem b_in_s), },
+        { intros b b_in_s,
+          have b_div_n, by exact divs b (finset.mem_insert_of_mem b_in_s),
+          have a_prime : prime a, { exact primes a (finset.mem_insert_self a s), },
+          have b_prime : prime b, { exact primes b (finset.mem_insert_of_mem b_in_s), },
+          rw k_times at b_div_n,
+          have r : b ∣ a ∨ b ∣ k, exact (prime.dvd_mul b_prime).mp b_div_n,
+          cases r with b_div_a b_div_k,
+          { exfalso,
+            have b_eq_a : b = a,
+            { cases (nat.dvd_prime a_prime).1 b_div_a with b_eq_1 b_eq_a,
+              { subst b_eq_1, exfalso, exact prime.ne_one b_prime rfl, },
+              { exact b_eq_a } },
+            subst b_eq_a,
+            exact a_not_in_s b_in_s, },
+          { exact b_div_k } } },
+    exact mul_dvd_mul_left a step, }
+end
+
+lemma primorial_le_pow_4 : ∀ (n : ℕ), n# ≤ 4 ^ n
+| 0 := le_refl _
+| 1 := le_of_inf_eq rfl
+| (n + 2) :=
+  match nat.mod_two_eq_zero_or_one (n + 1) with
+  | or.inl n_odd :=
+    match nat.even_iff.2 n_odd with
+    | ⟨m, twice_m⟩ :=
+      let recurse : m + 1 < n + 2 := by linarith in
+      by { calc (n + 2)#
```

Comment:
Please use `begin ... end` for multiline proofs.

Reply:
I'm afraid I need `recurse` to exist in term mode to satisfy the termination checker.

Reply:
That's fine: Johan just means don't use `by ...`.